NEXUS-5472: Cumulative WL fixes

cstamas opened the pull request:
Cumulative fixes gathered so far (mostly from RSO). 

Changes/Fixes:
- Repository#getLocalStatus proper handling (NEXUS-5591, NEXUS-5590)
- ProxyRepository#getProxyMode proper handling (NEXUS-5591)
- WL ThreadPool proper shutdown
- S3 scraper: did not neglect ".dot" files (NEXUS-5587)
- Apache HTTPD scraper: was too "tight" (NEXUS-5589)
- Added UTs (local status and proxy mode) and modified ITs to cover changes in NEXUS-5591

CI:
https://bamboo.zion.sonatype.com/browse/NXO-OSSF31-2
"passed" (with two unrelated failures)

CI:
https://bamboo.zion.sonatype.com/browse/NXO-OSSF31-3

cstamas committed:
NEXUS-5472: Events not propagating

Local status and proxy mode related events were not
getting to WL manager, as event dispatcher did prematurely
some checks (that were wrong and redundant).

Also, some slight method naming fix.

cstamas committed:
NEXUS-5472: Auto blocked proxies should become "frozen"

WL should not be removed nor update tried.

cstamas committed:
NEXUS-5472: UT verifying LocalStatus and WL interaction

cstamas committed:
NEXUS-5472: UT verifying proxy mode and WL interaction

cstamas committed:
NEXUS-5587: dot file filtering for S3 scraper

With UT. The S3 that is very much different from other
scrapers were missing filtering for dot files.

By simply adding a dot entry into UT, this added
coverage.

cstamas committed:
NEXUS-5589: HTTPD scraper "un-tighten"

The Apache httpd scraper was looking for address
element as mandatory requirement. This is wrong
as it depends on server configuration, looking
and checking it only if present, but it's non
presence does not affect final decision. If present
but have unexpected value, it will prevent scraping.

cstamas committed:
NEXUS-5591: Local mode and proxy mode handling corrected

Initially these two modes of a repo (and proxy repo) were
mishandled. This change now implements proper handling of
those and extends tests covering it.

cstamas committed:
NEXUS-5472: Kill pool if threads not end cleanly

Added shutdownNow call used in case of needed, when
on shutdown the running threads cannot be canceled.

In some cases, like remote HTTP request being "stuck",
like there was case on Bamboo and Codehaus snapshots
repo, where request times out (30s) or for some reason
takes 20s to execute, this might help if this
happens during shutdown (which is often in our ITs)

adreghiciu commented:
+1 (sanity checks only)

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/EventDispatcher.java:
```diff
@@ -100,7 +102,14 @@ protected void handleRepositoryAdded( final MavenRepository mavenRepository )
+        }
+        catch ( IllegalStateException e )
+        {
+            getLogger().info( e.getMessage() );
```
Why info log, and why just getMessage() ?  This could end up logging null if ISE was thrown w/o a message.

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/WLManagerImpl.java:
```diff
@@ -501,56 +522,75 @@ protected void updateAndPublishWhitelist( final MavenRepository mavenRepository,
+            else
+            {
+                getLogger().info( "Repository {} not supported by white-list feature, not updating it.",
+                    RepositoryStringUtils.getFullHumanizedNameString( mavenRepository ) );
```
Please stop

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/WLManagerImpl.java:
```diff
@@ -501,56 +522,75 @@ protected void updateAndPublishWhitelist( final MavenRepository mavenRepository,
+            {
+                if ( notify )
+                {
+                    getLogger().info( "Updated and published white-list of {}.",
```
putting

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/proxy/maven/wl/internal/WLManagerImpl.java:
```diff
@@ -501,56 +522,75 @@ protected void updateAndPublishWhitelist( final MavenRepository mavenRepository,
+            {
+                if ( notify )
+                {
+                    getLogger().info( "Unpublished white-list of {} (and is marked for noscrape).",
```
periods in log messages :-P

jdillon commented:
+0.95816

cstamas committed:
NEXUS-5472: Working on remaining 0.04184

Changes:
* removed declared but never thrown exceptions from WLManager#initializeWhitelist
* update and log cleanup of EventDispatcher
* removal of redundant method from WLManagerImpl
* building better world (for devs)
* targeting west coast without dots

cstamas committed:
NEXUS-5472: Removed redundant log

cstamas committed:
Lessen debug smog.

cstamas committed:
NEXUS-5472: WL IT helper plugin and client.

This was needed, as existing "wait for" solution was very unreliable, as
it was detecting only "transitions", but that proved not enough for
recent proxyMode changes.

Now, IT helper plugin exposes same method used by UT to "wait for
bg wl jobs", and client implemented using this new resource.

cstamas committed:
NEXUS-5472: ITs redirected to new resource and cluiebt

cstamas committed:
Missing license hdr

cstamas committed:
Merge remote-tracking branch 'origin/master' into nexus-5472-fixes-wip

cstamas commented:
CI +1'd
https://bamboo.zion.sonatype.com/browse/NXO-OSSF31-8

(since changes that introduced WL Test clients, only the Cargo related tests were sporadically fail, no WL related IT did fail)
